libnopegl/src/node_gblur.c

```diff
@@ -0,0 +1,543 @@
+    int dst_is_resizeable;
+    struct rendertarget_layout dst_layout;
+    struct rtt_ctx *dst_rtt_ctx;
+
+    struct gpu_block direction;
+    struct gpu_block kernel;
+    struct pgcraft *crafter;
+    struct pipeline_compat *pl_blur_h;
+    struct pipeline_compat *pl_blur_v;
+};
+
+#define OFFSET(x) offsetof(struct gblur_opts, x)
+static const struct node_param gblur_params[] = {
+    {"source",            NGLI_PARAM_TYPE_NODE, OFFSET(source),
+                          .node_types=(const uint32_t[]){NGL_NODE_TEXTURE2D, NGLI_NODE_NONE},
+                          .flags=NGLI_PARAM_FLAG_NON_NULL | NGLI_PARAM_FLAG_DOT_DISPLAY_FIELDNAME,
+                          .desc=NGLI_DOCSTRING("source to use for the blur")},
+    {"destination",       NGLI_PARAM_TYPE_NODE, OFFSET(destination),
+                          .node_types=(const uint32_t[]){NGL_NODE_TEXTURE2D, NGLI_NODE_NONE},
+                          .flags=NGLI_PARAM_FLAG_NON_NULL | NGLI_PARAM_FLAG_DOT_DISPLAY_FIELDNAME,
+                          .desc=NGLI_DOCSTRING("destination to use for the blur")},
+    {"bluriness",         NGLI_PARAM_TYPE_F32, OFFSET(bluriness_node), {.f32=0.03f},
+                          .flags=NGLI_PARAM_FLAG_ALLOW_NODE,
+                          .desc=NGLI_DOCSTRING("amount of bluriness in the range [0, 1]")},
+    {"max_kernel_size",   NGLI_PARAM_TYPE_U32, OFFSET(max_kernel_size_node), {.u32=127},
+                          .flags=NGLI_PARAM_FLAG_ALLOW_NODE,
+                          .desc=NGLI_DOCSTRING("maximum kernel size to be used internally, "
+                          "a higher value increase the maximum blur amount at the cost of "
+                          "performance, must be in the range [3;" CONSTANT_TO_STR(MAX_KERNEL_SIZE) "]")},
+    {NULL}
+};
+
+#define O(i) (2 * (i))
+#define W(i) (O(i) + 1)
+
+static int update_kernel(struct ngl_node *node)
+{
+    struct gblur_priv *s = node->priv_data;
+    struct gblur_opts *o = node->opts;
+
+    float bluriness = *(float *)ngli_node_get_data_ptr(o->bluriness_node, &o->bluriness);
+    if (bluriness < 0.0)
+        return NGL_ERROR_INVALID_ARG;
+
+    uint32_t max_kernel_size = *(uint32_t *)ngli_node_get_data_ptr(o->max_kernel_size_node, &o->max_kernel_size);
+    if (max_kernel_size < 3 || max_kernel_size > MAX_KERNEL_SIZE) {
+        LOG(ERROR, "maximum kernel size \"%u\" must be in the range [3;" CONSTANT_TO_STR(MAX_KERNEL_SIZE) "]", max_kernel_size);
+        return NGL_ERROR_INVALID_ARG;
+    }
+
+    if (s->bluriness == bluriness && s->max_kernel_size == max_kernel_size)
+        return 0;
+
+    s->bluriness = bluriness;
+
+    /* Round to the previous odd unsigned integer as the kernel size must be odd */
+    s->max_kernel_size = (max_kernel_size - 1) | 0x1;
+    s->max_radius_size = s->max_kernel_size - 1;
+
+    float radius_f = NGLI_CLAMP(bluriness, 0.f, 1.f) * (float)s->max_radius_size;
+    int32_t radius = NGLI_MIN((int32_t)ceil(radius_f), s->max_radius_size);
```

Comment:
nit: I would avoid calling functions within `NGLI_MIN`/`NGLI_MAX` macro due to the repetition; you may want to add an intermediate `const int32_t ceil_radius = (int32_t)ceil(radius_f)` intermediate

Reply:
``` diff
diff --git a/libnopegl/src/node_gblur.c b/libnopegl/src/node_gblur.c
index 91bc98f1..1150dd49 100644
--- a/libnopegl/src/node_gblur.c
+++ b/libnopegl/src/node_gblur.c
@@ -121,11 +121,11 @@ static int update_kernel(struct ngl_node *node)
     struct gblur_priv *s = node->priv_data;
     struct gblur_opts *o = node->opts;

-    float bluriness = *(float *)ngli_node_get_data_ptr(o->bluriness_node, &o->bluriness);
+    const float bluriness = *(float *)ngli_node_get_data_ptr(o->bluriness_node, &o->bluriness);
     if (bluriness < 0.0)
         return NGL_ERROR_INVALID_ARG;

-    uint32_t max_kernel_size = *(uint32_t *)ngli_node_get_data_ptr(o->max_kernel_size_node, &o->max_kernel_size);
+    const uint32_t max_kernel_size = *(uint32_t *)ngli_node_get_data_ptr(o->max_kernel_size_node, &o->max_kernel_size);
     if (max_kernel_size < 3 || max_kernel_size > MAX_KERNEL_SIZE) {
         LOG(ERROR, "maximum kernel size \"%u\" must be in the range [3," CONSTANT_TO_STR(MAX_KERNEL_SIZE) "]", max_kernel_size);
         return NGL_ERROR_INVALID_ARG;
@@ -140,8 +140,9 @@ static int update_kernel(struct ngl_node *node)
     s->max_kernel_size = (max_kernel_size - 1) | 0x1;
     s->max_radius_size = s->max_kernel_size - 1;

-    float radius_f = NGLI_CLAMP(bluriness, 0.f, 1.f) * (float)s->max_radius_size;
-    int32_t radius = NGLI_MIN((int32_t)ceil(radius_f), s->max_radius_size);
+    const float radius_f = NGLI_CLAMP(bluriness, 0.f, 1.f) * (float)s->max_radius_size;
+    const int32_t radius_i = (int32_t)ceil(radius_f);
+    const int32_t radius = NGLI_MIN(radius_i, s->max_radius_size);

     /*
      * Compute sigma for a given precision (1e-3f should be fine for up to
@@ -150,7 +151,7 @@ static int update_kernel(struct ngl_node *node)
      * - https://en.wikipedia.org/wiki/Talk%3AGaussian_blur#Radius_again
      * - https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule
      */
-    float sigma = (radius_f + 1.f) / sqrtf(-2.f * logf(1e-3f));
+    const float sigma = (radius_f + 1.f) / sqrtf(-2.f * logf(1e-3f));

     /*
      * Compute the weights for the interval [-radius, radius].
@@ -166,7 +167,7 @@ static int update_kernel(struct ngl_node *node)
     size_t nb_weights = 0;

     float sum = 0.0;
-    float sig = sigma * sqrtf(2.f);
+    const float sig = sigma * sqrtf(2.f);
     for (int i = -radius; i <= radius; i++) {
         const float p1 = erff(((float)i - 0.5f) / sig);
         const float p2 = erff(((float)i + 0.5f) / sig);
```